### ISSUE TITLE
Yielding from inside a lambda

### ISSUE THREAD

**Issue by ddanielou** [bobthecow/mustache.php#257]
Hi,

When using yield inside a lambda, shouldn't the lambda be treated as iterable instead of as a helper? I realize that, until it is invoked, such a lambda is not a generator yet, but it would be trivial to test if it is using [ReflectionFunction::isGenerator](https://php.net/manual/en/reflectionfunctionabstract.isgenerator.php). Not sure about performance though, or if that would be desirable at all.

Example:

    $m = new Mustache_Engine();
    echo $m->render('{{#foo}}{{.}}{{/foo}}', ['foo' => function(){
        yield 1;
        yield 2;
        yield 3;
    }]);

triggers

    strpos() expects parameter 1 to be string, object given

Workaround would be to `call_user_func($lambda_that_yields)`, but that somehow doesn't feel right…

**Comment by bobthecow** [comment]
Yeah, that's an interesting one. If you use a view model (or presenter) class, you can yield inside methods, and that already works.

Closures are an interesting case, because if the *value* is a closure, it is always treated as a lambda (higher order section) in Mustache. Letting you use a closure that's actually a generator would be nice, but it would also mean it's an exception to the "if the value is a closure" rule, adding even more potential confusion to an already confusing distinction.

**Comment by ddanielou** [comment]
The distinction is indeed confusing… having never used lambdas in templates before, I had to dig into the docs to understand why this didn't work readily. So yeah, I get your point.
But still: I can't think of a good reason to use yield inside a helper lambda. The whole point of generators is that you iterate over them, right? So you *could* assume that the least surprising thing would be to treat them the same as any old traversable object?

I made a quick patch, will look into properly forking your project tomorrow if you're interested.

**Comment by RomLAURENT** [comment]
HI,

But iterators are PHP 5.5 feature, as ReflectionFunction::isGenerator().

For now you could use a filter to do that :

```php

$template='{{# mygenerator | LambdaGet}}{{.}}{{/ mygenerator | LambdaGet}}';

$m= new Mustache_Engine([
    'pragmas'=>[Mustache_Engine::PRAGMA_FILTERS],
    'helpers'=>[
        'LambdaGet'=>function($lambda){
            return call_user_func($lambda);
        }
    ]
]);

echo $m->render($template,[
    'mygenerator'=>function (){
        yield 1;
        yield 2;
        yield 3;
    }
]);
```

This filter is used to get about the same meaning for lambda in section TAG as in interpolation TAG. it could help if your lambda return an object, bool, int, etc too.

**Comment by bobthecow** [comment]
To handle this case now, I'd probably use filters combined with a helper that forces iteration, rather than complicating the "is callable" / lambda logic.

[One such example is listed in the wiki](https://github.com/bobthecow/mustache.php/wiki/Helper-and-filter-cookbook#php-style-foreach).

**Comment by ddanielou** [comment]
Wow, thanks! I haven't used Mustache for a few years now and didn't even know it was still actively developed. I really appreciate seeing that kind of long-term support for "basically solved" tech.
